### PR TITLE
Facility access control

### PR DESCRIPTION
**Pull request by jeremyzitomer-usds**
## Related Issue or Background Info

- Rounds out https://app.zenhub.com/workspaces/simplereport-5ff8be8854ae8b000e378210/issues/cdcgov/prime-simplereport/581
- Added facility-based security restrictions to a bunch of methods
- Added a ton of new annotations in `AuthorizationConfiguration` and corresponding `UserAuthorizationVerifier` methods
- Converted a lot of String IDs in the graphql layer, to UUIDs
- gave the `addPatient`, `updatePatient`, and `addTestResult` mutations actual non-empty return types
- Added a facility filter on retrieving patients, and on retrieving a specific patient's test results
- Refactored @TomNUSDS 's facility permission function into a `@PreAuthorize` annotation
- Added `addFacilityAuthorities` and `removeFacilityAuthorities` methods in `TestUserIdentities`, and added `updateSelfPrivileges` in `BaseApiTest`
- A whole boatload of tests for facility access

### CODE REVIEW

File: backend/src/main/java/gov/cdc/usds/simplereport/api/patient/PatientMutationResolver.java
Comment (TomNUSDS) [inline backend/src/main/java/gov/cdc/usds/simplereport/api/patient/PatientMutationResolver.java:102]: We needed to go through and convert the Strings into UUIDs. Thanks.

File: backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java
Comment (TomNUSDS) [inline backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java:135]: The visually unmatched `+ ")"` bothers my mental parser.

Would almost prefer something like:
`@PreAuthorize("( " + SPEL_HAS_PERMISSION + "READ_RESULT_LIST" + ")" + " && " + SPEL_CAN_VIEW_TEST_EVENT)

Maybe it bother me because you're not doing `+ ") && " +` but breaking them out.

But it's not really an issue. And I guess not having `"(" +` keeps the lines shorter.

File: backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java
Comment (TomNUSDS) [inline backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java:131]: Nice. Thanks for the detailed docs.

File: backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java:135]: The parenthesis mismatch is my fault—there's an open paren in `SPEL_HAS_PERMISSION` because it's a function call up to the argument list. I stand by it, but it does make compound expressions naaasty to parse.

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java
Comment (jeremyzitomer-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:188]: felt like an important patch to sneak in

File: backend/src/test/java/gov/cdc/usds/simplereport/test_util/TestDataFactory.java
Comment (jeremyzitomer-usds) [inline backend/src/test/java/gov/cdc/usds/simplereport/test_util/TestDataFactory.java:66]: changed this so it doesn't conflict with config facilities if created as part of config org

**Comment by jeremyzitomer-usds** [comment]
only a couple api tests remaining

File: backend/src/test/resources/test-result-query
Comment (jeremyzitomer-usds) [inline backend/src/test/resources/test-result-query:1]: moved the old contents of this file to `test-results-query`

File: backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java
Comment (jeremyzitomer-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java:88]: @aliciabeckett-gov @katiealoisi @hmyers-cms would like your input here. are we sure, when loading the test results for a specific patient, we really only want to see the test results from the caller's accessible facilities? would the full history of the patient's test results constitute a medical history that the caller should have full access to when loading the patient?

**Comment by jeremyzitomer-usds** [comment]
@TomNUSDS @nullflux @benwarfield-usds @nickscialli-usds @jeskew-gov @timothybest-usds @timbest-usds this PR is ready for review. don't be intimidated by the size of it -- the vast majority of it is changing parameter types and return types, and a boat load of tests. but those aren't the most important parts.

The most important parts are:
- The new **annotations** in AuthorizationConfiguration and corresponding UserAuthorizationVerifier methods
- The new **facility-based security restrictions** on lots of service-level methods (specifically the annotations on those methods), listed in layman's terms below.
- A new **facility filter** on retrieving patients, and on retrieving a specific patient's test results

Facility-based restrictions on methods (note -- this is separate from role-based restrictions on these methods, which still apply):
- **addPatient**: caller must have access to the patient's facility; no restrictions if patient's facility is null
- **getPatient**: same as addPatient
- **getArchivedPatient**: same as addPatient
- **setPatientIsDeleted**: same as addPatient

- **updatePatient**: caller must have access to both the patient's original facility, and the new facility the patient is being assigned to; no additional restrictions if either of these facilities are null (e.g., if a patient is originally at Facility X and is updated to have a null facility, the caller only needs to have access to Facility X)

- **getPatients (no facility specified)**: will return only patients that are in facilities the caller can access, or whose facility is null
- **getPatients (facility specified)**: caller must have access to the specified facility; will return only patients that are in that facility or whose facility is null
- **getPatientsCount**: same as getPatients

- **addPatientToQueue**: caller must have access to the specified facility; the specified patient's facility must be null or the same as the specified facility

- **addTestResult**: caller must have access to the facility of the specified patient (or no additional restrictions if the patient's facility is null) _and_ the facility of the patient's unfinished test
- **updateTimeOfTestQuestions**: same as addTestResult
- **removePatientFromQueue**: same as addTestResult

- **getTestResult**: caller must have access to the facility of the specified test
- **editQueueItem**: same as getTestResult
- **correctTestMarkAsError**: same as getTestResult

- **getTestResults**: caller must have access to the facility of the specified patient (or no additional restrictions if the patient's facility is null); will return only tests that were performed in facilities that are accessible to the caller

- **getQueue**: caller must have access to the specified facility
- **getTestEventsResults**: same as getQueue

- **sendToPatientLink**: caller must have access to the facility of the test associated with the patient link
- **refreshPatientLink**: same as sendToPatientLink

**Comment by aliciabeckett-gov** [comment]
A few questions:

>A new facility filter on retrieving patients, and on retrieving a specific patient's test results

In the future, we may want to allow users to search for a patient at _all_ facilities, even if they don't have access to that facility. If we do end up wanting to do that, will this implementation make it hard to add or easy to add?

>addPatient(facility): caller must have access to the patient's facility; no restrictions if patient's facility is null

I'm assuming this is in the add patient UI. Will this limit the facility drop down menu in the UI to only the facilities the user has access to?

None of these should block the PR going out.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java
Comment (aliciabeckett-gov) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java:88]: IMO we should load the past test results for that patient, even if the user doesn't have access to the facility with the result. The caller should have access to the latest test results if they have access to the patient.

**Comment by jeremyzitomer-usds** [comment]
a) it'll require some tweaking, but not much! just rewiring a couple functions
b) that's partially a @nickscialli-usds question. i'm assuming the answer is yes because i imagine that drop down comes from the user's own accessible facilities?

**Comment by nickscialli-usds** [comment]
@jeremyzitomer-usds at this point any list of facilities comes from `organization.testingFacility`, which,  if I understand correctly, is only returning facilities a user has access to.

Do we have any facility list exposed by the backend that _isn't_ going to be restricted by the user's facility access? If the answer is no, then there will be no way for a user to see any facility they don't have access to in the UI.

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java
Comment (nickscialli-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:188]: 👀

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PersonService.java
Comment (nickscialli-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/PersonService.java:193]: Yes, I'd say it's an improvement when you no longer need the variable _actual_ Id

File: backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java
Comment (nickscialli-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java:88]: A question re: assigning users to facilities that I have, possibly because I wasn't privy to initial discussions on the data model:

Currently, when you create a patient, you select their facility (or specify "All Facilities). I can see how this is more performant (e.g., having fewer patients to search through when creating a test). But might it also cause confusion and/or data issues if a patient was created at Facility X and is now going for a test at Facility Y? A system user might think (a) this patient isn't in the system at all, so I should create a new record and (b) this patient has never been tested before.

I'm not sure how likely this scenario is at all, just a thought that crossed my mind.

**Review by nickscialli-usds — COMMENTED** [review]
Looks good (mostly from FE perspective)

File: backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java:88]: @aliciabeckett-gov to confirm, this function is to get all test results for a patient, not just the most recent. the most recent test will always be available to anyone who can see the patient. can you confirm that the caller should have access to _all_ previous tests of a patient (not just the most recent), regardless of facility?

Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java:88]: And @nickscialli-usds after this PR, a patient at Facility X indeed can't go for a test at Facility Y -- https://github.com/CDCgov/prime-simplereport/pull/1072/files#diff-ea6e46ce76b782b8c29dbe1ed4006220753b58756ba28063764398ca5f051007R167. I agree this could be a confusing user experience, though perhaps something to fix after this PR. Thoughts @hmyers-cms @aliciabeckett-gov @katiealoisi ?

**Comment by jeremyzitomer-usds** [comment]
> Do we have any facility list exposed by the backend that isn't going to be restricted by the user's facility access? If the answer is no, then there will be no way for a user to see any facility they don't have access to in the UI.

The answer is, in fact, no.

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java
Comment (jeskew-gov) [reply backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:188]: Not a blocker: Wouldn't suspended users already not be able to log in?

Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:188]: yes, but this is a method to retrieve other users in your organization. a user who has been deleted (SUSPENDED, in Okta) shouldn't come up in that user list

**Comment by benwarfield-usds** [comment]
Still working my way through, but one general question: do we have any sense of the before and after number of DB queries? It looks like it could be a bunch more, but I suspect that hibernate may optimize away some of those problems.

**Comment by jeremyzitomer-usds** [comment]
@benwarfield-usds not sure about "before" numbers, but here's the approximate additional number of queries that would be called for various API calls:

-submitTest/updateTest/updateToT: 7
-editPatient/removePatientFromQueue: 7
-patientSearch/getPatient/deletePatient: 4
-editQueueItem: 4
-getTestResult/getTestResults(for patient or facility): 3
-correctTestMarkAsError: 3
-refresh/sendPatientLink: 3
-addPatient: 3
-getQueue/addPatientToQueue: 3

the quickest way to reduce some of the higher numbers would be to factor out the redundant `isValidUser()` calls. other ones, i have a harder time figuring out how to avoid. that said, if hibernate does in fact optimize redundant queries, then then `isValidUser()` calls would make a smaller dent, so the bigger numbers above would in fact be a couple queries smaller.

**Comment by jeremyzitomer-usds** [comment]
@nickscialli-usds do these random github action failures look like the ones you got? only seeing them after merging from `main`

File: backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java
Comment (aliciabeckett-gov) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java:88]: @jeremyzitomer-usds good call out! The entry only should only have access to the most recent test result for the person being tested.

File: backend/src/test/java/gov/cdc/usds/simplereport/service/PersonServiceTest.java
Comment (benwarfield-usds) [inline backend/src/test/java/gov/cdc/usds/simplereport/service/PersonServiceTest.java:130]: Better practice would be to have different test case for the positive and negative tests—either using a tailored `@WithMockUser` or with custom annotations for the interesting cases.

File: backend/src/test/java/gov/cdc/usds/simplereport/service/PatientLinkServiceTest.java
Comment (benwarfield-usds) [inline backend/src/test/java/gov/cdc/usds/simplereport/service/PatientLinkServiceTest.java:39]: This pattern seems ugly to me—I know it's the natural extension of the way I handled a similar problem last week, which also seemed ugly to me, so I'm not throwing any shade at the doing of what needed to be done in order to get across the finish line, but we should probably figure out better patterns or we're going to have a lot of unmaintainable test code in the future.

**Review by benwarfield-usds — APPROVED** [review]
I have one substantial, albeit not necessarily serious, concern:

> addTestResult(patient): caller must have access to the facility of the specified patient (or no additional restrictions if the patient's facility is null) and the facility of the patient's unfinished test

I believe I commented on this in the code where I found it: this seems to create a lot of opportunities for tests to be started that require a special user to finish them. I think that's probably not desirable—if somebody's in the test queue, then they're in the test queue and can be operated on normally.

I also have qualms about extending the declarative approach to permission checking this far—it's the logical extension of how we've been doing things, and I would have had qualms about abandoning that, too, so I don't mean that as a critique of the decisions that were made in developing this, but it works out to be a pretty big chunk of permission-verifying code that examines a lot of the same entities. I think it can be made less sprawling with some judicious refactoring, which would help, but I think there's a risk to trying to make everything very explicit in this way, and we might want to consider a refactor of the service layer that puts the current facility list and the current organization both into the default filters of every query, so that we get the filtering we want implicitly rather than explicitly. That is in any case a conversation for another day, likely after the current crunch has passed.

**Comment by benwarfield-usds** [comment]
On the database-load question, I think it's probably the wrong thing to worry about _too_ much—though thanks for trying to rein it in where you could!—but I do worry about nickel-and-diming ourselves into less-than-great API responsiveness, and I especially worry because we still don't have any kind of proactive performance testing.

File: backend/src/test/java/gov/cdc/usds/simplereport/service/PatientLinkServiceTest.java
Comment (jeremyzitomer-usds) [reply backend/src/test/java/gov/cdc/usds/simplereport/service/PatientLinkServiceTest.java:39]: i had thought about doing something akin to runWith() that takes an executable, but then i ran into the problem of having to run a _callable_ with a particular set of facility authorities, then a _callable that returned multiple objects_, and ultimately ended up with the paradigm that i needed to be able to just add and remove facility authoriities. i'm open to doing something a bit "cleaner" like the `updateSelfPrivileges()` in the API tests which erases the need to have separate add and remove methods, but beyond that, i think we'd need to substantially reenvision our test scaffolding

File: backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java:88]: and what about a **_standard_** user? should a standard user with limited facility access still have access to all of a patient's past tests, regardless of facilities? (hoping selfishly that the the answer is no, because if the answer is yes i'd need to rewire things a bit.)

File: backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java:317]: i know SEARCH_PATIENTS isn't a perfect fit, but i think it's the best we have right now, unless we create a new VIEW_PATIENT permission. getPatient() is also used in the EditPatient panel, so START_TEST wouldn't make sense there.

File: backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/config/AuthorizationConfiguration.java]: > there was already a risk of that (patient has an open order at a facility you can't see)

yeah, there was this risk already, and this basically codifies it. in theory you should only be updating a test for a patient if you can see the facility their test was created at, but this gets to your review-level question that i've pushed to alicia katie and heather. good point.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java
Comment (aliciabeckett-gov) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java:88]: They should have access to a patient's _**last**_ test result, regardless of facilities for the time of test questions.

However if this is going to cause a big refactor we can skip it, log a ticket, and deal with it if we get a complaint.

Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java:88]: got it. that is how the current system is designed and won't require a refactor :) thanks!

File: backend/src/test/java/gov/cdc/usds/simplereport/service/PersonServiceTest.java
Comment (jeremyzitomer-usds) [reply backend/src/test/java/gov/cdc/usds/simplereport/service/PersonServiceTest.java:130]: wouldn't be able to use `@WithMockUser` since that requires a static declaration of authorities when I can't provide them (since the authorities need DB-generated facility UUIDs). otherwise noted.

File: backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java
Comment (jeremyzitomer-usds) [reply backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java:1013]: oops, this was a refactor and un-refactor mistake!

File: backend/src/main/java/gov/cdc/usds/simplereport/config/authorization/UserAuthorizationVerifier.java
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/config/authorization/UserAuthorizationVerifier.java:141]: wdym?

File: backend/src/main/java/gov/cdc/usds/simplereport/config/authorization/UserAuthorizationVerifier.java
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/config/authorization/UserAuthorizationVerifier.java:153]: will add `containsFacility`

File: backend/src/main/java/gov/cdc/usds/simplereport/config/authorization/UserAuthorizationVerifier.java
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/config/authorization/UserAuthorizationVerifier.java]: will factor it out

File: backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java:161]: > Checking if the patient and the facility are in the same org is redundant: we're already checking that they're both in the current org.

i'm not aware of any other code doing this

**Pull request merged**
